eZ/Publish/Core/Persistence/Legacy/URL/Query/CriterionHandler/VisibleOnly.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace eZ\Publish\Core\Persistence\Legacy\URL\Query\CriterionHandler;
+
+use eZ\Publish\API\Repository\Values\URL\Query\Criterion;
+use eZ\Publish\Core\Persistence\Legacy\URL\Query\CriteriaConverter;
+use eZ\Publish\Core\Persistence\Legacy\URL\Query\CriterionHandler;
+use eZ\Publish\Core\Persistence\Database\SelectQuery;
+use PDO;
+
+class VisibleOnly implements CriterionHandler
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function accept(Criterion $criterion)
+    {
+        return $criterion instanceof Criterion\VisibleOnly;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(CriteriaConverter $converter, SelectQuery $query, Criterion $criterion)
+    {
+        return $query->expr->in('ezurl.id', $this->getVisibleOnlySubQuery($query));
+    }
+
+    /**
+     * Generate query that selects ids of visible URLs.
+     *
+     * @param \eZ\Publish\Core\Persistence\Database\SelectQuery $query
+     * @return string
+     */
+    protected function getVisibleOnlySubQuery(SelectQuery $query)
+    {
+        // TODO: The following query requires optimization
+        $subSelect = $query->subSelect();
+        $subSelect
+            ->selectDistinct('ezurl_object_link.url_id')
```

Comment:
1. Querying is a responsibility of a Gateway, I'm not a big fan of injecting gateways in services other than persistence handler, but since it produces some sort of statement it should be obtained directly from the GW. BTW. I know we didn't follow everywhere this practices in the past ;-)
2. Please reuse Gateway's constants for table names.